100% code coverage and linting changes

KirillTregubov opened the pull request:
PR changes:
- Renamed `getErrrorMessage` -> `getErrorMessage`
- Added default value for `opts` instead of defining it after check
- Fixed `prefer-regex-literals` lint rule by removing the RegExp constructor
- Wrote tests for case when registering plugin without options according to expectations from stream
  - Every endpoint should have a response schema, unless explicitly disabled, aka, set to false
  - Every POST/PUT/PATCH should have a body, unless explicitly set to false

ATTENTION REQUIRED: unresolved issue with the current PR
On stream you said expected behaviour for the plugin is to enforce the `body` schema on POST, PUT, and PATCH routes. Currently, if `required: ["body"]` isn't passed in, it does not expect `body` schemas on the aforementioned routes. The test case on lines 48-64 in `test/enforce-schema.test.js` encodes this expectation.
Please let me know if this is undesired behaviour so I can fix it, or let me know to remove the related test.

Thanks in advance for your time!

KirillTregubov committed:
[FIX] - fix getErrorMessage naming, regex lint error, Test no options when registering

KirillTregubov committed:
ATTENTION REQUIRED: unclear behaviour in the third test

KirillTregubov committed:
fix - linting

agjs commented on test/enforce-schema.test.js:
```diff
@@ -32,19 +32,37 @@ test('Should pass if no options passed and schema validation explicitly disabled
   t.equal(res.payload, 'ok')
 })
 
+test('Should fail if no options passed: every POST should have a body unless explicitly set to false', async (t) => {
```
You are absolutely right, body and response are enforced by default, so one would need to do the opposite, **explicitly** set them to false.

KirillTregubov replied:
Should the `params` schema also be required by default on routes with params?

If this is the case, I don't quite understand the necessity for the `required` option, since to my understanding the default behaviour is to require all three of `schema`, `body`, and `params`. Am I missing something or should we remove this option?
If we can remove it, I propose to replace it with a new `disabled` option to disable specific schemas from being validated.

agjs replied:
* **body** required for PATCH, PUT and POST
* params required (as you said), on routes with params
* response required on everything

This PR is/was suppose to introduce an ability for users to *explicitly* disable any of the 3 types. So in simple words, things are enforced by default, and e.g. if you don't respond with anything in some endpoint, you also have to explicitly state that.

agjs replied:
So I think you are more or less right. On the schemas that I previously mentioned, you would want an ability to primarily disable some of the schemas.

KirillTregubov replied:
Another question: in `utils.js` there is an export called `initialExcludes` which excludes a bunch of docs subdirectories by default. This behaviour isn't documented anywhere, so I assume this was primarily for the project you originally built this for. Should I:
1. Document this behaviour in the README
2. Remove these excludes and give you the code required to exclude them in the required project(s) so you can upgrade to the new version

agjs replied:
Great question. Those excludes are specific to [Fastify Swagger](https://github.com/fastify/fastify-swagger) plugin. It excludes the routes created by Swagger, and my sentiment was that such side effect should be handled in here, rather then a consumer of this library needing to manually exclude something that they even aren't owning themselves.

agjs replied:
So in simple words, I think we should leave those in there without needing to document it, as documenting internal library behavior that has no impact on the user would not introduce any clarity, and it would not be useful in any way.

agjs replied:
```js
test('Should fail if body schema is missing', async (t) => {
  t.plan(1)

  const fastify = Fastify()

  await fastify.register(enforceSchema, { required: ['body'] })

  try {
    fastify.post('/foo', { schema: {} }, (req, reply) => {
      reply.code(201).send('ok')
    })
  } catch (error) {
    t.equal(error.message, 'POST: /foo is missing a body schema')
  }
})
```

If you are referring to this test, yes, this should fail, as you have provided a schema as an empty object literal.

KirillTregubov replied:
Will try to figure out what's causing the bug and post a commit by Monday.

Thanks so much for your elaboration, and Happy New Year to you and your family!

agjs replied:
> Will try to figure out what's causing the bug and post a commit by Monday.
> 
> Thanks so much for your elaboration, and Happy New Year to you and your family!

If anything, at the very least, I can thank you for putting effort at this level of detail into this. Whenever you are ready to merge, just let me know. Happy new year to you and your family as well. Looking forward to collaborate on various things in the upcoming year.

Cheers :) 🥳 🎈

KirillTregubov committed:
[FEAT]: deprecated required for disabled, added isSchemaDisabled util

Also updated README and abstracted tests to separate suites

KirillTregubov committed:
linting fixes

KirillTregubov commented:
> PR changes:
> 
> * Renamed `getErrrorMessage` -> `getErrorMessage`
> * Added default value for `opts` instead of defining it after check
> * Fixed `prefer-regex-literals` lint rule by removing the RegExp constructor
> * Wrote tests for case when registering plugin without options according to expectations from stream
>   
>   * Every endpoint should have a response schema, unless explicitly disabled, aka, set to false
>   * Every POST/PUT/PATCH should have a body, unless explicitly set to false
> 
> ATTENTION REQUIRED: unresolved issue with the current PR On stream you said expected behaviour for the plugin is to enforce the `body` schema on POST, PUT, and PATCH routes. Currently, if `required: ["body"]` isn't passed in, it does not expect `body` schemas on the aforementioned routes. The test case on lines 48-64 in `test/enforce-schema.test.js` encodes this expectation. Please let me know if this is undesired behaviour so I can fix it, or let me know to remove the related test.
> 
> Thanks in advance for your time!

Along with these changes, this PR now also:
- Deprecates `required` in favor of `disabled`
- updates README 
  - added badges
  - updated usage examples to reflect https://github.com/fastify/fastify/issues/4085#issuecomment-1165527150
  - updated options documentation
- added isSchemaDisabled util
- updated getErrorMessage to include messages for missing schema and other errors
- abstracted tests to separate files

ATTENTION STILL REQUIRED:
- my question [above](https://github.com/Programmer-Network/fastify-enforce-schema/pull/3#discussion_r1059538398) about `initialExcludes`
- in `tests/experimental.test.js`, the second test called `Should this fail since response schema invalid?` makes it seem like the schema is not being enforced beyond the HTTP response code. What is the expected behaviour here?

KirillTregubov committed:
fix - disable plugin early

KirillTregubov committed:
updated conflicting test to better show expected behaviour

KirillTregubov committed:
Merge remote-tracking branch 'upstream/main'

KirillTregubov committed:
fix - response checks and errors, changed hasProperties

Also moved experimental tests to main test file, cleaned up for merge

KirillTregubov commented:
PR is ready for review! We should definitely bump the version to 1.0.8 before merge.
Changes:
- All schema enforcement is enabled by default where appropriate
- Deprecates `required` in favor of `disabled`
- Updates README
  - added badges consistent with other fastify plugins
  - updated usage examples to reflect https://github.com/fastify/fastify/issues/4085#issuecomment-1165527150
  - updated options documentation
- Renamed `getErrrorMessage` -> `getErrorMessage`
- Fixed existing linting errors
- Reached 100% code coverage
- Changed the way the response schema is enforced, added support for `default` and `2xx`-like keys according to https://www.fastify.io/docs/latest/Reference/Validation-and-Serialization/#serialization
- Wrote tests for case when registering plugin without options according to expectated behaviours, tests for incomplete response schemas, and schemas with top level parameters, util changes
- added `isSchemaDisabled` util, tweaked `hasProperties` util
- updated getErrorMessage to include messages for missing schema and other errors
- abstracted util tests to separate file

KirillTregubov committed:
bump version to 1.0.8
